api/api/utils/moderation_lock.py

```diff
@@ -0,0 +1,122 @@
+import time
+
+import django_redis
+import structlog
+from redis import Redis
+from redis.exceptions import ConnectionError
+
+
+REPORT_LOCK_PREFIX = "soft_lock_report"
+MODERATOR_LOCK_PREFIX = "soft_lock_moderator"
+PREFIXES = [REPORT_LOCK_PREFIX, MODERATOR_LOCK_PREFIX]
+
+logger = structlog.get_logger(__name__)
+
+
+class LockManager:
+    def __init__(self, media_type):
+        self.media_type = media_type
+        redis: Redis = django_redis.get_redis_connection("default")
+        try:
+            redis.ping()
+            self.redis = redis
+        except ConnectionError:
+            logger.error("Redis connection failed")
+            self.redis = None
+
+    def prune(self):
+        """Delete all expired locks."""
+
+        def _prune(pattern: str):
+            now = int(time.time())
+            pipe = self.redis.pipeline()
+            for key in self.redis.keys(pattern):
+                for value, score in self.redis.zrange(key, 0, -1, withscores=True):
+                    if score <= now:
+                        logger.info("Deleting expired lock", key=key, value=value)
+                        pipe.zrem(key, value)
+            pipe.execute()
+
+        for prefix in PREFIXES:
+            _prune(f"{prefix}:*")
```

Comment:
Two issues here, one rather minor, and the other very much not minor.

First, the minor issue: is there a reason to avoid in-lining the closure's body as the loop body, rather than using a closure? The closure doesn't capture anything other than `self`, as far as I can tell, and using a loop would end up with virtually the same code, but in a more typical style. The utility of the closure here isn't obvious to me in this case, so it at least deserves an explanatory comment.

```suggestion
        pipe = self.redis.pipeline()
        for prefix in PREFIXES:
            pattern = f"{prefix}:*"
            now = int(time.time())
            for key in self.redis.keys(pattern):
                for value, score in self.redis.zrange(key, 0, -1, withscores=True):
                    if score <= now:
                        logger.info("Deleting expired lock", key=key, value=value)
                        pipe.zrem(key, value)

        pipe.execute()
```

That has the added (free) benefit of using the same pipeline, which is a small but nice improvement.

The other very significant issue, which is a blocker for this PR, is the use of `keys`, [which is O(n) where `n` is the number of keys in the database total (not just under the prefix)](https://redis.io/docs/latest/commands/keys/), and must not under any circumstances be used in our production code, as it will grind Redis to a halt. `keys` is Redis's version of a full-table scan.

Instead, we need to build a set of known keys and iterate through them. Luckily, the number of keys under the MODERATOR_LOCK_PREFIX is known, and easily retrievable:

```py
def get_moderator_locks(self):
    admin_usernames = get_user_model().objects.filter(...).values_list('username', flat=True)

    moderator_locked_works = redis.mget([f"{MODERATOR_LOCK_PREFIX}:{username}" for username in admin_usernames])

def prune():
    moderator_locks = self.get_moderator_locks()

    # do the pruning...
```

With this in mind, because we can never safely select all the `REPORT_LOCK_PREFIX`, I think we should ditch the report -> moderator mapping, and only use the moderator -> report mapping. The number of possible moderators is very small, so retrieving the set of locked works for each moderator is trivial (and literally infinitely less time-complex than the other way around).

Coupled with the polling approach I'm suggesting in the PR review comment, this should also simplify the overall approach.

Removing the report lock prefix does, strictly speaking, hurt reads when creating the table. However, the size of the table is limited, and realistically the search space of _locked works_ inside the list of works locked by each moderator will be very small, usually 0, and when not 0, then something like 1 and 2. It's trivial to create a set of those IDs and annotate the report list in Django ORM using that set at query time, and use it to represent the row. In fact, that's even _more_ efficient than the current approach of making a separate Redis request for each row of the table, like we are now.